File: 227. 基本计算器 II/main.java
class Solution {
    public int calculate(String s) {
        int n = s.length();
        List<String> l1 = new ArrayList<>();
        StringBuilder num = new StringBuilder();
        boolean flag = false;
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (c == ' ') {
            } else if (Character.isDigit(c)) {
                num.append(c);
                flag = true;
            } else {
                if (flag) {
                    l1.add(String.valueOf(num));
                    num = new StringBuilder();
                    flag = false;
                }
                l1.add(String.valueOf(c));
            }
        }
        if (!num.toString().equals("")) l1.add(num.toString());
        Stack<String> el = new Stack<>();
        Stack<String> op = new Stack<>();
        for (String ss : l1) {
            if (Character.isDigit(ss.charAt(0))) {
                el.push(ss);
            } else {
                if (op.isEmpty()) {
                    op.push(ss);
                } else {
                    while (!op.isEmpty() && !check(op.peek(), ss))
                        el.push(op.pop());
                    op.push(ss);
                }
            }
        }
        while (!op.isEmpty()) el.push(op.pop());
        l1.clear();
        while (!el.isEmpty()) l1.add(el.pop());
        Stack<Integer> stack = new Stack<>();
        n = l1.size();
        for (int i = n - 1; i >= 0; i--) {
            String ss = l1.get(i);
            if (Character.isDigit(ss.charAt(0))) {
                stack.push(Integer.valueOf(ss));
            } else {
                int r = stack.pop(), l = stack.pop();
                if (ss.equals("+")) {
                    stack.push(l + r);
                } else if (ss.equals("-")) {
                    stack.push(l - r);
                } else if (ss.equals("*")) {
                    stack.push(l * r);
                } else {
                    stack.push(l / r);
                }
            }
        }
        return stack.pop();
    }

    boolean check(String a, String b) {
        return (a.equals("+") || a.equals("-")) && (b.equals("*") || b.equals("/"));
    }
}

File: 227. 基本计算器 II/main.cc
class Solution {
private:
    stack<int> nums;
    stack<char> ops;
    map<char, int> priority;

public:
    int calculate(string s) {
        initialize_priority();
        s = regex_replace(s, regex(" "), "");
        s = regex_replace(s, regex("\\(-"), "(0-");
        nums.push(0);
        int i = 0;
        while (i < s.size()) {
            if (s[i] == ')') {
                while (!ops.empty()) {
                    if (ops.top() == '(') {
                        ops.pop();
                        break;
                    }
                    calc();
                }
                i++;
            } else if (s[i] == '(') {
                ops.push(s[i]);
                i++;
            } else {
                if (is_digit(s[i])) {
                    int num = 0;
                    while (is_digit(s[i])) {
                        num = num * 10 + (s[i] - '0');
                        i++;
                    }
                    nums.push(num);
                } else {
                    while (!ops.empty() && ops.top() != '(') {
                        if (priority[ops.top()] >= priority[s[i]]) {
                            calc();
                        } else {
                            break;
                        }
                    }
                    ops.push(s[i]);
                    i++;
                }
            }
        }
        while (!ops.empty()) {
            calc();
        }
        return nums.top();
    }

    void calc() {
        if (nums.size() < 2 || ops.empty()) {
            return ;
        }
        int num1 = nums.top(); nums.pop();
        int num2 = nums.top(); nums.pop();
        char op = ops.top(); ops.pop();
        int ans = 0;
        switch (op) {
            case '+':
                ans = num2 + num1;
                break;
            case '-':
                ans = num2 - num1;
                break;
            case '*':
                ans = num2 * num1;
                break;
            case '/':
                ans = num2 / num1;
                break;
            case '%':
                ans = num2 % num1;
                break;
            case '^':
                ans = pow(num2, num1);
        }
        nums.push(ans);
    }

    bool is_digit(char c) {
        return '0' <= c && c <= '9';
    }

    void initialize_priority() {
        priority['-'] = 1;
        priority['+'] = 1;
        priority['*'] = 2;
        priority['/'] = 2;
        priority['%'] = 2;
        priority['^'] = 3;
    }
};